KeenClient/KIODBStore.m

```diff
@@ -143,35 +157,45 @@ - (BOOL)deleteAndRecreateCorruptDB {
 }
 
 - (BOOL)openAndInitDB {
-    if(!dbIsOpen) {
-        if (![self openDB]) {
-            return false;
-        } else {
-            if(![self createTables]) {
-                KCLogError(@"Failed to create SQLite table!");
-                [self closeDB];
-                return false;
-            }
-        
-            if (![self migrateTable]) {
-                KCLogError(@"Failed to migrate SQLite table!");
-                [self closeDB];
-                return false;
-            }
-        
-            if (![self prepareAllSQLiteStatements]) {
+
+    // The database can be opened from the thread calling into the SDK,
+    // or by the database queue thread if it has been closed, so open
+    // needs to be protected with a lock.
+    [openLock lock];
```

Comment:
Yeah I thought we were eventually going to move away from this call happening so often. If that's the goal, and it seems realistic, then I wouldn't worry about optimizing here yet. Let's say all these issues get cleared up and we're still taking this lock thousands of times during normal course of operations...maybe then let's revisit turning this into a user-mode operation when not under contention (whether via atomics or slimmer locks or whatever).